packages/workbox-background-sync/lib/Queue.mjs

```diff
@@ -0,0 +1,297 @@
+/*
+ Copyright 2017 Google Inc. All Rights Reserved.
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+import {_private} from 'workbox-core';
+import StorableRequest from './StorableRequest.mjs';
+
+
+const {indexedDBHelper, WorkboxError} = _private;
+const names = new Set();
+
+
+const DB_NAME = 'workbox-background-sync';
+const TAG_PREFIX = 'workbox-background-sync';
+const OBJECT_STORE_NAME = 'requests';
+const MAX_RETENTION_TIME = 1000 * 60 * 60 * 24 * 7; // 7 days
+
+
+/**
+ * A class to manage storing failed requests in IndexedDB and retrying them
+ * later. All parts of the storing and replaying process are observable via
+ * callbacks.
+ *
+ * @example
+ * // Manually detecting failed requests and added them to the queue.
+ * const bgQueue = new workbox.backgroundSync.Queue('myQueue');
+ * self.addEventListener('fetch', function(event) {
+ *   if (!event.request.url.startsWith('https://example.com')) return;
+ *
+ *   const clone = event.request.clone();
+ *   event.respondWith(fetch(event.request).catch((err) => {
+ *     bgQueue.addRequest(clone);
+ *     throw err;
+ *   }));
+ * });
+ *
+ * @memberof module:workbox-background-sync
+ */
+export default class Queue {
+  /**
+   * Creates an instance of Queue with the given options
+   *
+   * @param {string} name The unique name for this queue. This name must be
+   *     unique as it's used to register sync events and store requests
+   *     in IndexedDB specific to this instance. An error will be thrown if
+   *     a duplicate name is detected.
+   * @param {Object} [param2]
+   * @param {number} [param2.maxRetentionTime = 7 days] The amount of time (in
+   *     ms) a request may be retried. After this amount of time has passed,
+   *     the request will be deleted and not retried.
```

Comment:
Nit .: reword to "deleted from the queue." insteand of "deleted and not retried."

Reply:
Done.